### DISCUSSION TITLE
$@ drops quotation marks

### DISCUSSION THREAD

**Discussion by auermich93** [git-for-windows/git#3385] [Q&A]
### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options
git version 2.31.1.windows.1
cpu: x86_64
built from commit: c5f0be26a7e3846e3b6268d1c6c4800d838c6bbb
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver
Microsoft Windows [Version 10.0.19043.1165]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt
Editor Option: VIM
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?
NO

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

I start the git shell within PowerShell.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
$ANDROID_HOME/build-tools/30.0.3/apksigner.bat sign --ks "C:\Program Files\Java\jdk1.8.0_131\jre\bin\KeyStore.jks" --ks-key-alias <my domain> --ks-pass pass:<my password> -v <some apk file>
```
Putting now above functionality into a shell script called `apksigner.sh`:

```
#! /bin/sh
apksigner=$ANDROID_HOME/build-tools/30.0.3/apksigner.bat
$apksigner $@
```

And calling finally the script with the same arguments: `./apksigner.sh sign --ks "C:\Program Files\Java\jdk1.8.0_131\jre\bin\KeyStore.jks" --ks-key-alias <my domain> --ks-pass pass:<my password> -v <some apk file>`.

 - What did you expect to occur after running these commands?

I expected that both commands lead to the same result.

 - What actually happened instead?

The first command works. However, outsourcing the functionality into a script fails, because the quotation marks of
the windows path (see parameter `--ks`) are dropped, leading to an invalid path. I could fix the problem by either
wrapping `$@` in quotation marks or passing the specific parameter in quotation marks, i.e. adjusting the last line of the script as follows: `$apksigner $1 $2 "$3" ...` (`$3` refers to the `--ks` argument)
Is this due to the automatic path resolving incorporated in git bash or why are the quotation marks dropped?

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

NO.

**Comment by dscho** [comment]
> `./apksigner.sh sign --ks "C:\Program Files\Java\jdk1.8.0_131\jre\bin\KeyStore.jks" --ks-key-alias <my domain> --ks-pass pass:<my password> -v <some apk file>`.

I am fairly certain that the quotes get interpreted as "this is one argument" (which in this case is totally unnecessary because the command-line argument does not contain any whitespace that would cause it to be mistaken for multiple arguments. That's what `/bin/sh` does with that command-line. You can see for yourself, more minimally by using an ad-hoc alias in lieu of a full shell script:

```console
$ git -c alias.auermich="!f() { for arg; do echo `$arg; done; }; f" auermich 1 2 3 "4 5"
1
2
3
4 5
```

It would get quite tricky if you wanted to preserve the double-quotes because multiple levels of quoting/unquoting are going on there, and different escape styles (Bash would want backslashes, while PowerShell uses the backtick as escape character).

However, I suspect that you want those double quotes in order to preserve the backslashes in the path? That would not work because Unix shells interpret backslashes as escape characters even within double-quoted arguments, you would have to single-quote it to fix it. But then, you could simply pass on the command-line arguments, _preserving_ them rather than letting the shell interpolate them. In your use case, it would probably suffice to surround the `$@` by double-quotes:

```sh
#! /bin/sh
apksigner=$ANDROID_HOME/build-tools/30.0.3/apksigner.bat
$apksigner "$@"
```
